tss-esapi/tests/context_tests/tpm_commands/duplication_commands_tests.rs

```diff
@@ -1,2 +1,131 @@
 // Copyright 2021 Contributors to the Parsec project.
 // SPDX-License-Identifier: Apache-2.0
+mod test_duplicate {
+    use crate::common::{create_ctx_with_session, create_ctx_without_session, decryption_key_pub};
+    use std::convert::TryFrom;
+    use tss_esapi::attributes::SessionAttributesBuilder;
+    use tss_esapi::constants::{tss::TPM2_CC_Duplicate, SessionType};
+    use tss_esapi::handles::{ObjectHandle, PersistentTpmHandle, TpmHandle};
+    use tss_esapi::interface_types::dynamic_handles::Persistent;
+    use tss_esapi::structures::{SymmetricDefinition, SymmetricDefinitionObject};
+    use tss_esapi::{
+        interface_types::{
+            algorithm::HashingAlgorithm,
+            resource_handles::{Hierarchy, Provision},
+            session_handles::PolicySession,
+        },
+        structures::Auth,
+    };
+
+    #[test]
+    fn test_duplicate() {
+        let mut context = create_ctx_with_session();
+
+        let random_digest = context.get_random(16).unwrap();
+        let key_auth = Auth::try_from(random_digest.value().to_vec()).unwrap();
+
+        let object_to_duplicate_handle: ObjectHandle = context
```

Comment:
I do not think you can duplicate a Primary key. Because duplicate is used to move an object from one Hierarchy to another (potentially on another machine)